proposals/infra/INF-0004-validator-hashing.md

```diff
@@ -195,6 +163,93 @@ multiple-year-old preview compiler for generating final compiled shaders for a
 title. The high rate of bugs in preview compilers contributes to making this
 even less likely.
 
+## Detailed Design
+
+### D3D Runtime Behavior
+
+Starting with AgilitySDK version TBD, the runtime adopts new behavior for how 
+and when the shader hash is validated, including support for bytecode validation 
+in the debug layer.
+
+For shaders with the `PREVIEW_BYPASS` hash, the shader is allowed to execute 
+only if developer mode and the experimental feature 
+`D3D12ExperimentalShaderModels` is enabled, otherwise the runtime will produce
+an error.
+
+For shaders with the `BYPASS` hash, shaders can run without validating the hash 
+regardless of whether the machine is in developer mode as long as the shader 
+targets a supported non-experimental shader model.
+
+All other values of the hash are assumed to be real hash values.
+
+For shaders that contain a real hash value, the behavior is unchanged. The hash
+is verified and the shader executes as it does with existing versions of the 
+runtime.
+
+> Note: an important behavior change here is that shaders with zero'd hash data
+> will always be treated as invalid by the runtime and rejected.
+> 
+> An exception is for a period of time already existing preview shaders that used 
+> zero'd hash data will be treated like having the `PREVIEW_BYPASS` hash in that
+> preview runtimes will allow them to be used, at least until those preview shaders
+> and runtimes are no longer supported.  In contrast to `PREVIEW_BYPASS` shaders,
+> these legacy zero'd has shaders will not run have the validator invoked on them
+> by the runtime/debug layer since that wasn't expected of them.
+
+#### D3D Debug Layer Validation Control
+
+Apps can configure how the debug layer runtime validates bytecode passed to it. 
+
+The validation is the same code the compiler uses when it endorses bytecode at 
+compile time by applying a hash.  A copy of this validation code is in 
+the debug layer for it to use.  So it would typically be 
+redundant to validate again, but it can be useful to validate shaders that have 
+the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 
+
+If the debug layer finds errors when running the bytecode validator it only 
+reports messages, but doesn't fail shader creation.
```

Comment:
Checking here that we don't want to enable a mode where shader creation fails if it fails validation?

Reply:
This is what the runtime validation option I had was for. The debug layer / runtime interaction is typically it is the runtime that does the failing, and the debug layer's role is just an observer giving commentary.

Gets messy to have to make sure all codepaths work cleanly if the debug layer suddenly starts failing on behalf of the runtime.

I'd be fine bringing back the option for the runtime to fail, as long as debug layer is on.

If we bring this back, I'd switch the enum back to a flags field, so the fail on error can be layered on.

Jesse's argument is an app can write 20 lines of code if they want, to intercept the debug errors and fail if they want.  Certainly an unusual burden to put on them.

Reply:
Or I guess they could just configure the debug layer to break on that error message as well.

I thought I'd run into cases where enabling the debug layer would cause some API calls to fail that didn't previously.

Reply:
Yes, we try to avoid situations where having the debug layer on changes behavior.  So I started here by proposing that you could ask for the runtime to fail independent of whether the debug layer is on or not.    But based on everyone's hesitation against having the runtime to validation backed off.

Yes, folks can configure the debug layer to break on the error.  Not quite as easy as it seems because there are tons of different codepaths where validation will be invoked from, and we'd have to try to guarantee that they always use the same error ID so folks can break on all possible instances of the failure.  There are some code patterns in the debug layer where there are error ID name specific to a shader stage, for instance, and those code paths would have to make an exception for this particular error.

Reply:
I will try to do this when I implement.  If when the dust settles I'm confident it's just one ID to break on, I can update this spec to list that ID.